### PR TITLE
Global note area

### LINKED ISSUE #83: Note area in Calender View
There is a note area where text can be posted. The contents can be be edited. The changes appear for every user. Colors can be added to notes (red:important, green:not important).

### PR DESCRIPTION
**Pull request by MarcoLeipnitz**
Front- and backend implementation of the global note area.
Notes are shared between all users and updated via SSE.

closes #83 

---

### Implementation

- [x] Does this code change accomplish what it is supposed to do?
- [x] Can this solution be simplified without unreasonable effort
- [x] Is the code modular enough?
- [x] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [x] Single Responsibility Principle: Does any class serve only one purpose? If not, it should probably split up in multiple classes.
- [x] Interface Segregation: If interface implemented, is it fully implemented?
- [x] Dependency Injection: Are central object instantiation processes managed by a single context?
- [x] If any functionality is missing, is it marked with a comment that starts with TODO:?
- [x] Do you see any potential to improve the performance of the code significantly? 

### Logic Errors and Bugs

- [x] Can you think of any use case in which the code does not behave as intended?
- [x] Can you think of any inputs or external events that could break the code?
- [x] Are edge cases covered by tests?

### Error Handling and Logging

- [x] Is error handling done the correct way?
- [x] Should any logging or debugging information be added or removed?
- [x] Are error messages user-friendly?

### Dependencies (internal and external)

- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Is the API.md up-to-date?
- [ ] Are the enumerations between frontend and backend synchronized?

### Security and Data Privacy

- [ ] Does the code introduce any security vulnerabilities?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is sensitive data like user data handled and stored securely?
- [ ] Is the right encryption used?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Were any secrets committed?

### Testing and Testability (later)

- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Do the existing tests reasonably cover the code change?
- [ ] Are there some test cases, input or edge cases that should be tested in addition?

### Usability and Accessibility

- [ ] Is the proposed solution (UI) accessible (e.g. colorblindness, ...)?
- [ ] Is the API/UI intuitive to use?
- [ ] Was the app tested on another device?

### Readability

- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by smaller methods/different names?
- [ ] Is the code located in the right file/folder?
- [ ] Do you think certain methods should be restructured to have a more intuitive control flow?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message better?
- [ ] Could some comments be removed by making the code itself more readable?

### CODE REVIEW

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/GlobalNoteController.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        if(note.hasInvalidColor())
+            throw new InvalidColorParameterException("Color does not match 7 char HTML notation");
+
+        GlobalNote savedNote = globalNoteRepository.save(note);
+
+        URI location = ServletUriComponentsBuilder
+                .fromCurrentRequest()
+                .path("/{id}")
+                .buildAndExpand(savedNote.getId())
+                .toUri();
```
Comment (JohaJung) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/GlobalNoteController.java]: I like this pattern.

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/GlobalNoteController.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    @PostMapping
+    ResponseEntity<GlobalNote> createGlobalNote(@RequestBody GlobalNote note){
+        if(note.hasInvalidTitle())
+            throw new InvalidParameterException("Note titles may not be empty");
+
+        if(note.hasInvalidColor())
+            throw new InvalidColorParameterException("Color does not match 7 char HTML notation");
+
+        GlobalNote savedNote = globalNoteRepository.save(note);
+
```
Comment (JohaJung) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/GlobalNoteController.java]: This misses an SSE event (or TODO).

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/GlobalNoteController.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                .fromCurrentRequest()
+                .path("/{id}")
+                .buildAndExpand(savedNote.getId())
+                .toUri();
+
+        return ResponseEntity.created(location).body(savedNote);
+    }
+
+    @PutMapping("/{id}")
+    ResponseEntity<GlobalNote> updateGlobalNote(@PathVariable Long id, @RequestBody GlobalNote proposedNote){
```
Comment (JohaJung) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/GlobalNoteController.java]: This should check if there

* is an ID mismatch between URL and Body
* the ID really exists

and error otherwise.

See how I did at the calendar entities

--- 

this function misses SSE triggers, too (same below).

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/GlobalNoteController.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.servlet.support.ServletUriComponentsBuilder;
+
+import java.net.URI;
+import java.util.List;
+
+@AllArgsConstructor
+@RestController
+@RequestMapping("/globalNotes")
+public class GlobalNoteController {
```
[the remark this answers is not in the document]
Comment (MarcoLeipnitz) [reply VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/GlobalNoteController.java]: When I remove this annotation and replace it like in the Lecture controller I get a lot of metadata when I request that url.
I just want the JSON data, so that the frontend has less trouble with managing that as well

**Comment by JohaJung** [comment]
> This was by far not ready for review, but I appreciate the feedback

Joa mir bewusst, aber wenn ich eh schon durchgucke kann ich auch gleich kommentieren ^^

File: VLA-Frontend/src/components/globalNotes/GlobalNoteEntry.tsx [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+        }
+                
+
+        {
+          isEditing &&
+                <>
+                  <br/>
+                  <Button>
+                    {{
+                      text: "Notiz unwiderrufbar löschen",
```
[the remark this answers is not in the document]
Comment (MarcoLeipnitz) [reply VLA-Frontend/src/components/globalNotes/GlobalNoteEntry.tsx]: Was passiert, wenn man aus Versehen auf den Knopf kommt?
Ich möchte da keine weiteren Dialog einbauen.

Wenn man auf Editieren und dann löschen drückt, ist es sehr viel wahrscheinlicher, dass das so gewollt war.

**Review by JohaJung — APPROVED** [review]
The design is a bit kunterbunt, but should be okay for now.

**Review by CarolinRoell — APPROVED** [review]
